# Activity 2
### Measuring distance
In this activity you will measure a distance between two points on a map using different earth models, including flat Earth model, and sphere Earth model. Furthermore you will measure direction between the points.

#### 1, Measuring distance using the euclidean distance approach (i.e., using Pythagorean theorem), which denotes the flat Earth model.  

In this example, we'll measure the distance between Albury and Canberra using the Universal Transverse
Mercator (UTM) projection and decimal degrees (unprojected). The units of the UTM projection are in meters. The first point, defined as (x1,y1), represents Albury ( a city in New South Wales, Australia), while the second point, defined as (x2,y2)
represents Canberra the capital city of Australia, as shown in the following illustration:

In [2]:
import math

In [3]:
# The point for Albury
x1= 493258.55
y1= 6006785.06

In [4]:
# The point for Canberra
x2= 697686.97
y2= 6091664.87

In [5]:
# X distance, and Y distance
x_dist=x1-x2
y_dist=y1-y2

In [42]:
# Pythagorean theorem 
dist_sq = x_dist**2 + y_dist**2
distance = math.sqrt(dist_sq)
print(distance)
# 221349.41 metres

221349.40941717595


In [43]:
# convert from m to km
distance=distance/1000.0 

print(distance)

221.34940941717596


When measuring distance using decimal degrees, we must convert the angles into radians, which account for the curved surface distance between the coordinates. We'll multiply our output in radians by the radius of the Earth in meters to convert back from radians.

In [55]:
import math

# Albury point 
x1 = 146.92
y1 = -36.08

# Canberra point
x2 = 149.15 
y2 = -35.29

# computation
x_dist = math.radians(x1 - x2)
y_dist = math.radians(y1 - y2)
dist_sq = x_dist**2 + y_dist**2
dist_rad = math.sqrt(dist_sq)
dist_rad= dist_rad * 6371251.46
print(dist_rad)

263075.1139366397


In [56]:
# convert to km
dist_rad=dist_rad/1000.0
print(dist_rad)

263.0751139366397


In this method, we came up with around 263 kilometers, which is 42 kilometers more
than our first measurement. So, as you can see, your choice of measurement algorithm and
Earth model can have significant consequences. Using the same equation, we come up with
radically different answers, depending on our choice of coordinate system and Earth
model.

#### 2, The sphere Earth model using the haversine formula (please look this up online for further reading)
The haversine formula uses trigonometry to calculate Great Circle distance using coordinates defined in decimal degrees as input. The haversine formula is haversine(θ) = sin²(θ/2), where θ is the central angle between two points on a sphere.The
haversine formula is the most commonly used distance measuring formula because it is relatively lightweight from a coding perspective and reasonably accurate in most cases. It is considered to be accurate to within about a meter. In this method, we'll convert the axis distances from degrees into radians before we apply the formula, just like in the previous example. But this time, we'll also convert the latitude (yaxis) coordinates into radians separately.

In [59]:
import math
x1 = 146.92
y1 = -36.08
x2 = 149.15 
y2 = -35.29
x_dist = math.radians(x1 - x2)
y_dist = math.radians(y1 - y2)
y1_rad = math.radians(y1)
y2_rad = math.radians(y2)
a = math.sin(y_dist/2)**2 + math.sin(x_dist/2)**2 * math.cos(y1_rad) * math.cos(y2_rad)
c = 2 * math.asin(math.sqrt(a))
distance = c * 6371 # kilometers
print(distance)
# 219.72

219.72147973894042


#### Bearing
In addition to distance, you will often want to know the bearing of a line between its endpoints. We can calculate this line direction from one of the points using only the Python math module.

In [61]:
from math import atan2, cos, sin, degrees

In [62]:
# set up variables for our two points
lon1 = 146.92
lat1 = -36.08
lon2= 149.15 
lat2= -35.29

In [63]:
# calculate the angle between the two points
angle = atan2(cos(lat1)*sin(lat2)-sin(lat1) * \
cos(lat2)*cos(lon2-lon1), sin(lon2-lon1)*cos(lat2))


In [65]:
# calculate the bearing of the line in degrees
bearing = (degrees(angle) + 360) % 360
print(bearing)

# 219.655

219.65454464476434


Sometimes, you end up with a negative bearing value. To avoid this issue, we add 360 to the result to avoid a negative number and use the Python modulo operator to keep the value from climbing to over 360.

### End of activity